docs/design/2022-12-06-support-late-materialization.md

```diff
@@ -0,0 +1,144 @@
+
+## Objective
+- Improve performance of some AP queries.
+- Without performance degradation of most queries.
+
+## Design
+
+In brief, we will divide the filter conditions of selection(whose child is tablescan) into two parts: 
+- Those light and with high selectivity filter conditions, which will be pushed down to tablescan.
+- The rest filter conditions, which will be executed after tablescan.
+
+### TiDB side
+
+We will decide which filter conditions will be pushed down to tablescan in TiDB planner. We should consider the following factors:
+- Only the filter conditions with high selectivity should be pushed down.
+- The filter conditions which contain heavy cost functions should not be pushed down.
+- Filter conditions that apply to the same column are either pushed down or not pushed down at all.
+- The pushed down filter conditions should not contain too many columns.
+
+The algorithm is:
+```go
+func selectPushDownConditions(conditions []expression.Expression)([]expression.Expression) {
+    // Group them by the column, sort the conditions by the selectivity of the group.
+    // input: [cond1_on_col1, cond2_on_col2, cond3_on_col2, cond4_on_col1, cond5_on_col3]
+    // - group: [[cond1_on_col1, cond4_on_col1], [cond2_on_col2, cond3_on_col2], [cond5_on_col3]]
+    // - sort: [[cond2_on_col2, cond3_on_col2], [cond1_on_col1, cond4_on_col1], [cond5_on_col3]]
+    //   where group on col2 has the highest selectivity, and group on col1 has the second highest selectivity.
+    // output: [[cond2_on_col2, cond3_on_col2], [cond1_on_col1, cond4_on_col1], [cond5_on_col3]]
+    sorted_conds = groupByColumnSortBySelectivity(conditions)
+    for _, cond_group := range sorted_conds {
+        // If the group contains heavy cost functions, skip it.
+        // heavy cost functions: json functions, ...(to be discussed)
+        if withHeavyCostFunction(cond_group) {
+            continue
+        }
+        // If the group contains too many columns, skip it.
+        // too many columns: more than 25% of total number of columns to be read, ...(to be discussed)
+        if isTooManyColumns(cond_group) {
+            continue
+        }
+        // If the selectivity of the group is larger than the threshold, push down it.
+        // threshold: 0.1, ...(to be discussed)
+        if selectivity(cond_group) > threshold {
+            return cond_group
+        }
+    }
+    return nil
+}
+```
+
+Since we need statistics to calculate the selectivity of the filter conditions, so this algorithm should be executed in postOptimize phase.
+
+Obviously, beacuse the selectivity of the filter conditions is accurate, and the algorithm is not optimal, we can not guarantee that the pushed down filter conditions are the best. In order to patch a workaround for performance degradation, we will add a variable `/*+ set_var(enable_late_materialization=false) */` to disable optimizer to push down filter conditions.
```

Comment:
The set_var hint is not a GA hint ( It is developed by community and no document).
So if you want to set var, please 'set xxx=xxx'

Reply:
get, thanks